File: native/surface.c
#include <stdlib.h>
#include <cairo/cairo.h>
#include <cairo/cairo-pdf.h>
#include <cairo/cairo-ps.h>
#include <cairo/cairo-svg.h>
#include <cairo/cairo-script.h>

#include "argument.h"
#include "error.h"
#include "bind.h"
#include "factory.h"
#include "surface.h"
#include "device.h"
#include "list.h"


static cairo_user_data_key_t surfaceKey;

void surface_destroy(void* handle) {
    if (handle) {
        cairo_surface_t* surface = (cairo_surface_t*) handle;
        unsigned int refCount = cairo_surface_get_reference_count(surface);
        if (surface && refCount == 1) {
            cairo_surface_destroy(surface);
            surface = NULL;
        }

    }
}

void surface_data_destroy(void* handle) {
    if (handle)
        free(handle);
}

void surface_status(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*)bind_get_self(args);

    cairo_status_t status = cairo_surface_status(surface);

    Dart_Handle ret = factory_create_status(status);

    Dart_SetReturnValue(args, ret);
    Dart_ExitScope();
}


void image_surface_create(Dart_NativeArguments args) {
    Dart_EnterScope();
    Dart_Handle obj = arg_get(&args, 0);
    cairo_format_t format = (cairo_format_t)arg_get_int(&args, 1);
    int64_t width = arg_get_int(&args, 2);
    int64_t height = arg_get_int(&args, 3);

    cairo_surface_t* surface = cairo_image_surface_create(format, width, height);

    bind_setup(surface, obj, surface_destroy);

    Dart_SetReturnValue(args, Dart_Null());
    Dart_ExitScope();
}

void image_surface_create_for_data(Dart_NativeArguments args) {
    Dart_EnterScope();
    Dart_Handle obj = arg_get(&args, 0);
    Dart_Handle bytesList = arg_get(&args, 1);
    cairo_format_t format = (cairo_format_t)arg_get_int(&args, 2);
    int64_t width = arg_get_int(&args, 3);
    int64_t height = arg_get_int(&args, 4);
    int64_t stride = arg_get_int(&args, 5);

    int length = list_length(bytesList);

    unsigned char* arr = (unsigned char*)malloc(sizeof(unsigned char) * length);
    int i;
    for (i = 0; i < length; i++) {
        unsigned char b = (unsigned char)list_int_at(bytesList, i);
        arr[i] = b;
    }

    cairo_surface_t* surface = cairo_image_surface_create_for_data(arr, format, width, height, stride);

    cairo_surface_set_user_data(surface, &surfaceKey, arr, surface_data_destroy);

    bind_setup(surface, obj, surface_destroy);

    Dart_SetReturnValue(args, Dart_Null());
    Dart_ExitScope();
}

void image_surface_get_data(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*)bind_get_self(args);

    unsigned char* data = cairo_image_surface_get_data(surface);

    Dart_Handle result = Dart_Null();

    if (data) {
        int length = sizeof(data) / sizeof(data[0]);
        result = Dart_NewList(length);
        int i;
        for (i = 0; i < length; i++) {
            Dart_ListSetAt(result, i, Dart_NewInteger(data[i]));
        }
    }

    Dart_SetReturnValue(args, result);
    Dart_ExitScope();
}

void pdf_surface_create(Dart_NativeArguments args) {
    Dart_EnterScope();
    Dart_Handle obj = arg_get(&args, 0);
    Dart_Handle fileNameObj = arg_get(&args, 1);
    const char* fileName = NULL;
    if (!Dart_IsNull(fileNameObj)) {
        fileName = arg_get_string(&args, 1);
    }
    double width = arg_get_double(&args, 2);
    double height = arg_get_double(&args, 3);

    cairo_surface_t* surface = cairo_pdf_surface_create(fileName, width, height);

    bind_setup(surface, obj, surface_destroy);

    Dart_SetReturnValue(args, Dart_Null());
    Dart_ExitScope();
}

void pdf_surface_set_size(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*) bind_get_self(args);
    double width = arg_get_double(&args, 1);
    double height = arg_get_double(&args, 2);

    cairo_pdf_surface_set_size(surface, width, height);

    Dart_SetReturnValue(args, Dart_Null());
    Dart_ExitScope();
}

void pdf_surface_restrict_to_version(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*) bind_get_self(args);
    cairo_pdf_version_t version = (cairo_pdf_version_t) arg_get(&args, 1);

    cairo_pdf_surface_restrict_to_version(surface, version);

    Dart_SetReturnValue(args, Dart_Null());
    Dart_ExitScope();
}

void pdf_surface_get_versions(Dart_NativeArguments args) {
    Dart_EnterScope();
    const cairo_pdf_version_t* versions;
    int length;

    cairo_pdf_get_versions(&versions, &length);

    Dart_Handle list = Dart_NewList(length);

    int i;
    for (i = 0; i < length; i++) {
        Dart_ListSetAt(list, i, factory_create_pdf_version(versions[i]));
    }

    Dart_SetReturnValue(args, list);
    Dart_ExitScope();
}

void pdf_version_to_string(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_pdf_version_t version = (cairo_pdf_version_t)arg_get_int(&args, 1);

    const char* ver = cairo_pdf_version_to_string(version);

    Dart_Handle res = Dart_NewStringFromCString(ver);

    Dart_SetReturnValue(args, res);
    Dart_ExitScope();
}

void image_surface_get_width(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*) bind_get_self(args);

    int width = cairo_image_surface_get_width(surface);

    Dart_SetReturnValue(args, Dart_NewInteger(width));
    Dart_ExitScope();
}

void image_surface_get_height(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*) bind_get_self(args);

    int height = cairo_image_surface_get_height(surface);

    Dart_SetReturnValue(args, Dart_NewInteger(height));
    Dart_ExitScope();
}

void image_surface_get_stride(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*) bind_get_self(args);

    int stride = cairo_image_surface_get_stride(surface);

    Dart_SetReturnValue(args, Dart_NewInteger(stride));
    Dart_ExitScope();
}

void surface_finish(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*) bind_get_self(args);

    cairo_surface_finish(surface);

    Dart_SetReturnValue(args, Dart_Null());
    Dart_ExitScope();
}

void surface_flush(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*) bind_get_self(args);

    cairo_surface_flush(surface);

    Dart_SetReturnValue(args, Dart_Null());
    Dart_ExitScope();
}

void surface_get_content(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*) bind_get_self(args);

    int content = (int)cairo_surface_get_content(surface);

    Dart_SetReturnValue(args, Dart_NewInteger(content));
    Dart_ExitScope();
}

void surface_mark_dirty(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*) bind_get_self(args);

    cairo_surface_mark_dirty(surface);

    Dart_SetReturnValue(args, Dart_Null());
    Dart_ExitScope();
}

void surface_mark_dirty_rectangle(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*) bind_get_self(args);
    int x = arg_get_int(&args, 1);
    int y = arg_get_int(&args, 2);
    int width = arg_get_int(&args, 3);
    int height = arg_get_int(&args, 4);

    cairo_surface_mark_dirty_rectangle(surface, x, y, width, height);

    Dart_SetReturnValue(args, Dart_Null());
    Dart_ExitScope();
}

void surface_get_device_offset(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*) bind_get_self(args);
    double x;
    double y;

    cairo_surface_get_device_offset(surface, &x, &y);

    Dart_Handle point = factory_create_point(x, y);

    Dart_SetReturnValue(args, point);
    Dart_ExitScope();
}


void surface_set_device_offset(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*) bind_get_self(args);
    double x = arg_get_double(&args, 1);
    double y = arg_get_double(&args, 2);

    cairo_surface_set_device_offset(surface, x, y);

    Dart_SetReturnValue(args, Dart_Null());
    Dart_ExitScope();
}


void surface_copy_page(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*) bind_get_self(args);

    cairo_surface_copy_page(surface);

    Dart_SetReturnValue(args, Dart_Null());
    Dart_ExitScope();
}

void surface_show_page(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*) bind_get_self(args);

    cairo_surface_show_page(surface);

    Dart_SetReturnValue(args, Dart_Null());
    Dart_ExitScope();
}

void surface_supports_mime_type(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*) bind_get_self(args);
    const char* mime = arg_get_string(&args, 1);

    cairo_bool_t supports = cairo_surface_supports_mime_type(surface, mime);

    Dart_SetReturnValue(args, Dart_NewBoolean(supports != 0));
    Dart_ExitScope();
}


void surface_has_show_text_glyphs(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*) bind_get_self(args);

    cairo_bool_t res = cairo_surface_has_show_text_glyphs(surface);

    Dart_SetReturnValue(args, Dart_NewBoolean(res != 0));
    Dart_ExitScope();
}

void surface_get_type(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*) bind_get_self(args);

    int type = (int)cairo_surface_get_type(surface);

    Dart_SetReturnValue(args, Dart_NewInteger(type));
    Dart_ExitScope();
}


void surface_get_fallback_resolution(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*) bind_get_self(args);
    double xRes;
    double yRes;

    cairo_surface_get_fallback_resolution(surface, &xRes, &yRes);

    Dart_Handle resolutionObj = factory_create_resolution(xRes, yRes);

    Dart_SetReturnValue(args, resolutionObj);
    Dart_ExitScope();
}

void surface_set_fallback_resolution(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*) bind_get_self(args);
    double xRes = arg_get_double(&args, 1);
    double yRes = arg_get_double(&args, 2);

    cairo_surface_set_fallback_resolution(surface, xRes, yRes);

    Dart_SetReturnValue(args, Dart_Null());
    Dart_ExitScope();
}

void surface_get_device(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*) bind_get_self(args);

    cairo_device_t* device = cairo_surface_get_device(surface);

    Dart_Handle result = Dart_Null();
    if (device != NULL) {
        result = factory_create_object("_Device", "", NULL, 0);
        bind_setup(device, result, device_destroy);
    }

    Dart_SetReturnValue(args, result);
    Dart_ExitScope();
}

void image_surface_create_from_png(Dart_NativeArguments args) {
    Dart_EnterScope();
    Dart_Handle surfaceObj = arg_get(&args, 0);
    const char* fileName = arg_get_string(&args, 1);

    cairo_surface_t* surface = cairo_image_surface_create_from_png(fileName);

    bind_setup(surface, surfaceObj, surface_destroy);

    Dart_SetReturnValue(args, Dart_Null());
    Dart_ExitScope();
}

void surface_write_to_png(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*) bind_get_self(args);
    const char* fileName = arg_get_string(&args, 1);

    cairo_status_t status = cairo_surface_write_to_png(surface, fileName);

    error_verify(status);

    Dart_SetReturnValue(args, Dart_Null());
    Dart_ExitScope();
}

void ps_surface_create(Dart_NativeArguments args) {
    Dart_EnterScope();
    Dart_Handle obj = arg_get(&args, 0);
    Dart_Handle fileNameObj = arg_get(&args, 1);
    const char* fileName = NULL;
    if (!Dart_IsNull(fileNameObj)) {
        fileName = arg_get_string(&args, 1);
    }
    double width = arg_get_double(&args, 2);
    double height = arg_get_double(&args, 3);

    cairo_surface_t* surface = cairo_ps_surface_create(fileName, width, height);

    bind_setup(surface, obj, surface_destroy);

    Dart_SetReturnValue(args, Dart_Null());
    Dart_ExitScope();
}

void ps_level_to_string(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_ps_level_t level = (cairo_ps_level_t) arg_get_int(&args, 1);

    const char* str = cairo_ps_level_to_string(level);

    Dart_SetReturnValue(args, Dart_NewStringFromCString(str));
    Dart_ExitScope();
}

void ps_surface_set_eps(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*) bind_get_self(args);
    bool eps = arg_get_bool(&args, 1);

    cairo_ps_surface_set_eps(surface, eps);

    Dart_SetReturnValue(args, Dart_Null());
    Dart_ExitScope();
}

void ps_surface_get_eps(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*) bind_get_self(args);

    cairo_bool_t eps = cairo_ps_surface_get_eps(surface);

    Dart_SetReturnValue(args, Dart_NewBoolean(eps != 0));
    Dart_ExitScope();
}

void ps_surface_restrict_to_level(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*) bind_get_self(args);
    cairo_ps_level_t level = (cairo_ps_level_t) arg_get_int(&args, 1);

    cairo_ps_surface_restrict_to_level(surface, level);

    Dart_SetReturnValue(args, Dart_Null());
    Dart_ExitScope();
}

void ps_get_levels(Dart_NativeArguments args) {
    Dart_EnterScope();
    const cairo_ps_level_t* levels;
    int length = 0;

    cairo_ps_get_levels(&levels, &length);

    Dart_Handle list = Dart_NewList(length);

    int i;
    for (i = 0; i < length; i++) {
        Dart_ListSetAt(list, i, factory_create_ps_level(levels[i]));
    }

    Dart_SetReturnValue(args, list);
    Dart_ExitScope();
}

void ps_surface_set_size(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*)bind_get_self(args);
    double width = arg_get_double(&args, 1);
    double height = arg_get_double(&args, 2);

    cairo_ps_surface_set_size(surface, width, height);

    Dart_SetReturnValue(args, Dart_Null());
    Dart_ExitScope();
}

void ps_surface_dsc_begin_setup(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*)bind_get_self(args);

    cairo_ps_surface_dsc_begin_setup(surface);

    Dart_SetReturnValue(args, Dart_Null());
    Dart_ExitScope();
}

void ps_surface_dsc_begin_page_setup(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*)bind_get_self(args);

    cairo_ps_surface_dsc_begin_page_setup(surface);

    Dart_SetReturnValue(args, Dart_Null());
    Dart_ExitScope();
}

void ps_surface_dsc_comment(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*)bind_get_self(args);
    const char* comment = arg_get_string(&args, 1);

    cairo_ps_surface_dsc_comment(surface, comment);

    Dart_SetReturnValue(args, Dart_Null());
    Dart_ExitScope();
}

void svg_surface_create(Dart_NativeArguments args) {
    Dart_EnterScope();
    Dart_Handle obj = arg_get(&args, 0);
    Dart_Handle fileNameObj = arg_get(&args, 1);
    const char* fileName = NULL;
    if (!Dart_IsNull(fileNameObj)) {
        fileName = arg_get_string(&args, 1);
    }
    double width = arg_get_double(&args, 2);
    double height = arg_get_double(&args, 3);

    cairo_surface_t* surface = cairo_svg_surface_create(fileName, width, height);

    bind_setup(surface, obj, surface_destroy);

    Dart_SetReturnValue(args, Dart_Null());
    Dart_ExitScope();
}

void svg_version_to_string(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_svg_version_t version = (cairo_svg_version_t) arg_get_int(&args, 1);

    const char* str = cairo_svg_version_to_string(version);

    Dart_SetReturnValue(args, Dart_NewStringFromCString(str));
    Dart_ExitScope();
}

void svg_surface_restrict_to_version(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*)bind_get_self(args);
    cairo_svg_version_t version = (cairo_svg_version_t) arg_get_int(&args, 1);

    cairo_svg_surface_restrict_to_version(surface, version);

    Dart_SetReturnValue(args, Dart_Null());
    Dart_ExitScope();
}

void svg_get_versions(Dart_NativeArguments args) {
    Dart_EnterScope();
    const cairo_svg_version_t* versions;
    int length = 0;

    cairo_svg_get_versions(&versions, &length);

    Dart_Handle list = Dart_NewList(length);

    int i;

    for (i = 0; i < length; i++) {
        Dart_ListSetAt(list, i, factory_create_svg_version(versions[i]));
    }

    Dart_SetReturnValue(args, list);
    Dart_ExitScope();
}

void recording_surface_create(Dart_NativeArguments args) {
    Dart_EnterScope();
    Dart_Handle obj = arg_get(&args, 0);
    cairo_content_t content = (cairo_content_t) arg_get_int(&args, 1);
    Dart_Handle rectList = arg_get(&args, 2);

    cairo_surface_t* surface;

    if (!Dart_IsNull(rectList)) {
        int length = list_length(rectList);
        cairo_rectangle_t rects[length];
        int i;
        for (i = 0; i < length; i++) {
            Dart_Handle rect = list_at(rectList, i);
            cairo_rectangle_t r;
            r.x = list_double_at(rect, 0);
            r.y = list_double_at(rect, 1);
            r.width = list_double_at(rect, 2);
            r.height = list_double_at(rect, 3);
            rects[i] = r;
        }

        surface = cairo_recording_surface_create(content, rects);
    } else {
        surface = cairo_recording_surface_create(content, NULL);
    }

    bind_setup(surface, obj, surface_destroy);

    Dart_SetReturnValue(args, Dart_Null());
    Dart_ExitScope();
}

void recording_surface_ink_extents(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*)bind_get_self(args);
    double x = 0.0;
    double y = 0.0;
    double width = 0.0;
    double height = 0.0;

    cairo_recording_surface_ink_extents(surface, &x, &y, &width, &height);

    Dart_Handle result = factory_create_rectangle_double(x, y, width, height);

    Dart_SetReturnValue(args, result);
    Dart_ExitScope();
}

void recording_surface_get_extents(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*)bind_get_self(args);
    Dart_Handle rectList = arg_get(&args, 1);

    int length = list_length(rectList);
    cairo_rectangle_t rects[length];
    int i;
    for (i = 0; i < length; i++) {
        Dart_Handle rect = list_at(rectList, i);
        cairo_rectangle_t r;
        r.x = list_double_at(rect, 0);
        r.y = list_double_at(rect, 1);
        r.width = list_double_at(rect, 2);
        r.height = list_double_at(rect, 3);
        rects[i] = r;
    }

    cairo_bool_t bound = cairo_recording_surface_get_extents(surface, rects);


    Dart_SetReturnValue(args, Dart_NewBoolean(bound != 0));
    Dart_ExitScope();
}

void surfaces_equals(Dart_NativeArguments args) {
    Dart_EnterScope();
    cairo_surface_t* surface = (cairo_surface_t*) bind_get_self(args);
    Dart_Handle otherObj = arg_get(&args, 1);
    cairo_surface_t* other = (cairo_surface_t*) bind_get(otherObj);

    bool equals = surface == other;
    Dart_SetReturnValue(args, Dart_NewBoolean(equals));
    Dart_ExitScope();
}

void script_surface_create(Dart_NativeArguments args) {
    Dart_EnterScope();
    Dart_Handle obj = arg_get(&args, 0);
    cairo_device_t* device = (cairo_device_t*) bind_get(arg_get(&args, 1));
    cairo_content_t content = (cairo_content_t) arg_get_int(&args, 2);
    double width = arg_get_double(&args, 3);
    double height = arg_get_double(&args, 4);

    cairo_surface_t* surface = cairo_script_surface_create(device, content, width, height);

    bind_setup(surface, obj, device_destroy);

    Dart_SetReturnValue(args, Dart_Null());
    Dart_ExitScope();
}

void script_surface_create_for_target(Dart_NativeArguments args) {
    Dart_EnterScope();
    Dart_Handle obj = arg_get(&args, 0);
    cairo_device_t* device = (cairo_device_t*) bind_get(arg_get(&args, 1));
    cairo_surface_t* target = (cairo_surface_t*) bind_get(arg_get(&args, 2));

    cairo_surface_t* surface = cairo_script_surface_create_for_target(device, target);

    bind_setup(surface, obj, device_destroy);

    Dart_SetReturnValue(args, Dart_Null());
    Dart_ExitScope();
}

File: cairodart/lib/src/surface.dart
part of cairodart.base;

abstract class Surface {

  void finish();
  void flush();
  void markDirty();
  void markDirtyRect(int x, int y, int width, int height);
  void copyPage();
  void showPage();
  bool supportsMimeType(String mimeType);

  Content get content;
  Point get deviceOffset;
  void set deviceOffset(Point offset);
  bool get hasShowTextGlyphs;
  SurfaceType get surfaceType;
  Resolution get fallbackResolution;
  void set fallbackResolution(Resolution resolution);
  void writeTo(String fileName);
  Device get device;

  operator==(Surface other);
}

abstract class PdfSurface implements Surface {

  factory PdfSurface(String fileName, num widthInPoints, num heightInPoints) =>
    new _PdfSurface(fileName, widthInPoints, heightInPoints);

  void setSize(num widthInPoints, num heightInPoints);

  void restrictToVersion(PdfVersion version);

  List<PdfVersion> get versions;

}

abstract class PostScriptSurface implements Surface {

  factory PostScriptSurface(String fileName, num widthInPoints, num heightInPoints) =>
    new _PostScriptSurface(fileName, widthInPoints, heightInPoints);

  bool encapsulated;

  void restrictToLevel(PostScriptLevel level);

  List<PostScriptLevel> get levels;

  void setSize(num width, num height);

  void beginPageSetup();

  void beginSetup();

  void dscComment(String comment);

}

abstract class SvgSurface {

  factory SvgSurface(String fileName, num widthInPoints, num heightInPoints) =>
    new _SvgSurface(fileName, widthInPoints, heightInPoints);

  void restrictToVersion(SvgVersion version);

  List<SvgVersion> get versions;

}

abstract class RecordingSurface implements Surface {

  factory RecordingSurface(Content content, List<Rectangle> extents) => new _RecordingSurface(content, extents);

  Rectangle get inkExtents;

  bool getExtents(List<Rectangle> extents);

}

abstract class _Surface extends NativeFieldWrapperClass2 implements Surface {

  void finish() native 'surface_finish';
  void flush() native 'surface_flush';
  Content get content {
    int c = _getContent();
    return new _Content(c);
  }

  SurfaceType get surfaceType {
    int type = _getSurfaceType();
    return new _SurfaceType(type);
  }

  void markDirty() native 'surface_mark_dirty';
  void markDirtyRect(int x, int y, int width, int height) native 'surface_mark_dirty_rectangle';

  int _getContent() native 'surface_get_content';
  int _getSurfaceType() native 'surface_get_type';

  Point get deviceOffset native 'surface_get_device_offset';
  void set deviceOffset(Point offset) {
    _deviceOffset(offset.x, offset.y);
  }

  void _deviceOffset(double x, double y) native 'surface_set_device_offset';

  void copyPage() native 'surface_copy_page';
  void showPage() native 'surface_show_page';
  bool get hasShowTextGlyphs native 'surface_has_show_text_glyphs';
  bool supportsMimeType(String mimeType) native 'surface_supports_mime_type';

  Resolution get fallbackResolution native 'surface_get_fallback_resolution';
  void set fallbackResolution(Resolution resolution) {
    _fallbackResolution(resolution.xResolution, resolution.yResolution);
  }

  void _fallbackResolution(double x, double y) native 'surface_set_fallback_resolution';

  void writeTo(String fileName) native 'surface_write_to_png';

  Device get device native 'surface_get_device';

  operator==(Surface other) native 'surfaces_equals';
}

abstract class ImageSurface implements Surface {

  factory ImageSurface(Format format, int width, int height) => new _ImageSurface(format, width, height);
  factory ImageSurface.forData(List<int> data, Format format, int width, int height, int stride) =>
    new _ImageSurface.forData(data, format, width, height, stride);

  factory ImageSurface.fromPng(String fileName) => new _ImageSurface.fromPng(fileName);

  int get width;
  int get height;
  int get stride;
  Format get format;
  List<int> get data;

  void write();


}

abstract class ScriptSurface implements Surface {

  factory ScriptSurface(Device script, Content content, num width, num height) => new _ScriptSurface(script, content, width, height);
  factory ScriptSurface.forTarget(Device script, Surface target) => new _ScriptSurface(script, target);

}

class _ImageSurface extends _Surface implements ImageSurface {

  Format _format;

  _ImageSurface(Format format, int width, int height) {
    _createImageSurface(format.value, width, height);
    _format = format;
  }

  _ImageSurface.forData(List<int> data, Format format, int width, int height, int stride) {
    _createImageSurfaceForData(data, format.value, width, height, stride);
  }

  _ImageSurface.internal(){}

  void _createImageSurfaceForData(List<int> data, int format, int width, int height, int stride) native 'image_surface_create_for_data';

  void _createImageSurface(int format, int width, int height) native 'image_surface_create';

  int get width native 'image_surface_get_width';
  int get height native 'image_surface_get_height';
  int get stride native 'image_surface_get_stride';
  Format get format => _format;
  List<int> get data native 'image_surface_get_data';

  String _fileName;

  _ImageSurface.fromPng(this._fileName) {
    _createPngSurface(_fileName);
  }

  void _createPngSurface(String fileName) native 'image_surface_create_from_png';

  void write() => writeTo(this._fileName);


}

class _PdfSurface extends _Surface implements PdfSurface {

  _PdfSurface(String fileName, num widthInPoints, num heightInPoints) {
    _createPdfSurface(fileName, widthInPoints.toDouble(), heightInPoints.toDouble());
  }

  void _createPdfSurface(String fileName, double widthInPoints, double heightInPoints) native 'pdf_surface_create';

  void setSize(num widthInPoints, num heightInPoints) {
    _setSize(widthInPoints.toDouble(), heightInPoints.toDouble());
  }

  void _setSize(double width, double height) native 'pdf_surface_set_size';

  void restrictToVersion(PdfVersion version) {
    _restrictToVersion(version.value);
  }

  void _restrictToVersion(int version) native 'pdf_surface_restrict_to_version';

  List<PdfVersion> get versions native 'pdf_surface_get_versions';

}

class _PostScriptSurface extends _Surface implements PostScriptSurface {

  _PostScriptSurface(String fileName, num widthInPoints, num heightInPoints) {
    _createPostScriptSurface(fileName, widthInPoints.toDouble(), heightInPoints.toDouble());
  }

  void _createPostScriptSurface(String fileName, double width, double height) native 'ps_surface_create';

  bool get encapsulated native 'ps_surface_get_eps';
  void set encapsulated(bool value) native 'ps_surface_set_eps';

  void restrictToLevel(PostScriptLevel level) {
    _restrictToLevel(level.value);
  }

  void _restrictToLevel(int level) native 'ps_surface_restrict_to_level';

  List<PostScriptLevel> get levels native 'ps_get_levels';

  void setSize(num width, num height) {
    _setSize(width.toDouble(), height.toDouble());
  }

  void _setSize(double width, double height) native 'ps_surface_set_size';

  void beginPageSetup() native 'ps_surface_dsc_begin_page_setup';

  void beginSetup() native 'ps_surface_dsc_begin_setup';

  void dscComment(String comment) native 'ps_surface_dsc_comment';

}

class _SvgSurface extends _Surface implements SvgSurface {

  _SvgSurface(String fileName, num widthInPoints, num heightInPoints) {
    _createSvgSurface(fileName, widthInPoints.toDouble(), heightInPoints.toDouble());
  }

  void _createSvgSurface(String fileName, double width, double height) native 'svg_surface_create';

  void restrictToVersion(SvgVersion version) {
    _restrictToVersion(version.value);
  }

  void _restrictToVersion(int version) native 'svg_surface_restrict_to_version';

  List<SvgVersion> get versions native 'svg_get_versions';

}

class _RecordingSurface extends _Surface implements RecordingSurface {

  List<List<double>> extentsPoints(List<Rectangle> extents) => extents.map((r) => <double>[r.x.toDouble(), r.y.toDouble(), r.width.toDouble(), r.height.toDouble()]).toList();

  _RecordingSurface(Content content, List<Rectangle> extents) {
    if (extents != null) {
      _createRecordingSurface(content.value, extentsPoints(extents));
    } else {
      _createRecordingSurface(content.value, null);
    }
  }

  void _createRecordingSurface(int content, List<List<double>> rect) native 'recording_surface_create';

  Rectangle get inkExtents native 'recording_surface_ink_extents';

  bool getExtents(List<Rectangle> extents) => _getExtents(extentsPoints(extents));

  bool _getExtents(List<List<double>> rect) native 'recording_surface_get_extents';

}

class _ScriptSurface extends _Surface implements ScriptSurface {

  _ScriptSurface(Device script, Content content, num width, num height) {
    _createScriptSurface(script, content.value, width.toDouble(), height.toDouble());
  }

  _ScriptSurface.forTarget(Device script, Surface target) {
    _createScriptSurfaceForTarget(script, target);
  }

  void _createScriptSurface(Device script, int content, double width, double height) native 'script_surface_create';
  void _createScriptSurfaceForTarget(Device script, Surface target) native 'script_surface_create_for_target';

}